### DISCUSSION TITLE
Options for role or policy based authentication

### DISCUSSION THREAD

**Discussion by minorityuk** [oqtane/oqtane.framework#2308] [General]
Hi

I have an app currently in mvc using standard identity authentication and roles.  Original developer built the system where client has 1 or more databases and just quite messy.

Our product basically has 2 tiers of customer.  The customer we sell solution too and then that customer manages their clients on it.

The application needs the below structure where we have a tenant per customer and then they can have a sub tenant per their client.  Then clients have multiple locations so I thought they should be sub tenants too. Example is below.

-My customer 1
-----client A
--------location A
---------location B
-------client B
-----------location C

-my client 2
-------client c
----------location d

Does oqtane have the functionality to for enabling role or policy authentication per tenant for example:
- Customer would have 3 role tenant admin ( full access), edit access to all sub tenants, read only

- each user of client (let use client A) would need to be granted access to each location with option of the follow role
        - edit access , ability to edit items
        - read only 
        - manager ,view reports

If I wanted each client in a separate tenant database and then have separate customer tenant is there anyway to expose the client database data to customer to search.
The reason I ask is our customer provides products to clients on loan. If the customer gets a request from client b for a product then it should be possible to search all location of client a and b to see if its I  use and if not transfer it to the other location.

My thoughts are no without merging data or using something like elastic search?

**Comment by sbwalker** [comment]
Ok, so you have Customers... and Customers manage Clients... and Clients can have multiple Locations. What is not clear to me from your description is which of these entities have requirements to keep their data physically separated? 

The terminology in Oqtane is a bit confusing but the base unit for organizing entities is referred to as a Site, and a physical database used for storing information is called a Tenant. Oqtane supports 2 forms of multi-tenancy - shared and isolated. Isolated essentially means that a Site has its own Tenant (ie. physical database) which is separated from other Sites. Shared allows you to manage multiple Sites within the same Tenant (ie. Sites are logically separated). These concepts can be used in conjunction with one another - ie. Customers could be isolated, Clients shared, and Locations isolated. 

The most important thing to understand is that each Site is independent - there is no relationship between Sites regardless of whether they are in a shared or isolated Tenant. The exception to this rule is User data. In a shared tenancy model, the Users are shared between the various Sites which exist in the same Tenant (ie for single sign-on capabilities). However information such as security roles, files, modules, etc... are not shared.

So to answer your question about creating an aggregated search index across different Sites/Tenants, you are correct that you would need to create your own custom solution. You may be able to create an Oqtane scheduled job which crawls the various Sites in the installation and build your own index.
